src/visibility_tracker.rs

```diff
@@ -606,198 +595,204 @@ fn register_name<'a>(
 /// Collect all public items that are reachable from the crate root and record their parent Ids.
 fn visit_root_reachable_public_items<'a>(
     crate_: &'a Crate,
-    parents: &mut HashMap<&'a Id, HashSet<&'a Id>>,
     traversal_state: &NameResolution<'a>,
-    currently_visited_items: &mut HashSet<&'a Id>,
     item: &'a Item,
-    parent_id: Option<&'a Id>,
-) {
-    match item.visibility {
-        Visibility::Crate | Visibility::Restricted { .. } => {
-            // This item is not public, so we don't need to process it.
-            return;
-        }
-        Visibility::Public => {} // Public item, keep going.
-        Visibility::Default => {
-            // Enum variants, and some impls and methods have default visibility:
-            // they are visible only if the type to which they belong is visible.
-            // However, we don't recurse into non-public items with this function, so
-            // reachable items with default visibility must be public.
-        }
-    }
-
-    let item_parents = parents.entry(&item.id).or_default();
-    if let Some(parent_id) = parent_id {
-        item_parents.insert(parent_id);
+) -> HashMap<&'a Id, Vec<&'a Id>> {
+    #[derive(Debug)]
+    enum Action<'a> {
```

Comment:
It would be good to leave a doc comment explaining the purpose of this type.

It took me a non-trivial amount of time to figure out why it's present and how it works. I'm not confident that in 6 months you or I will still remember the details, so as a favor to our future selves, I'd like to err on the side of explaining in too much detail rather than too little.

Reply:
Fixed (by removing the refactor)